### PR TITLE
support windows

### LINKED ISSUE #21: Support Windows
We used UNIX-like operating system syscall[1] to get time for getting better performance, which caused EG can't be compiled in Windows. As EG has been opening source, we should support Windows-compatible.

[1] https://github.com/megaease/easegress/blob/c71016788bc7a44cf85745cc74429714463a5f29/pkg/common/time.go#L27

### PR DESCRIPTION
**Pull request by tg123**
fix #21 

signal idea copy from nginx <https://github.com/nginx/nginx/blob/a64190933e06758d50eea926e6a55974645096fd/src/os/win32/ngx_process_cycle.c>

`./easegress-server --signal-upgrade` introduced 

- [x] the graceful is not support on windows, will follow the pattern of nginx using Windows service to impl it
- [x] gh test on windows
- [x] signal test 
- [ ] gracenet to support windows

### CODE REVIEW

**Review by megaeasex — CHANGES_REQUESTED** [review]
[TASK:easegress-pr-test FAILED]megaease/easegress Pull Request 74 Deploy Test failed

**Review by megaeasex — CHANGES_REQUESTED** [review]
[TASK:easegress-pr-test FAILED]megaease/easegress Pull Request 74 Deploy Test failed

**Comment by tg123** [comment]
could this get merged to make windows supported in mainstream?
gracenet like cannot impl on windows at the moment.

**Review by megaeasex — CHANGES_REQUESTED** [review]
[TASK:easegress-pr-test FAILED]megaease/easegress Pull Request 74 Deploy Test failed

**Comment by benja-wu** [comment]
> could this get merged to make windows supported in mainstream?
> gracenet like cannot impl on windows at the moment.

Hi tg123, first of all, thank you for providing a Windows supporting PR for Easegress. 

1. It seems that after v1.0.1 and v1.1.0, the Easegress' code implementation has evolved a lot, so the megaeasex bot can't build it in the test environment successfully. So I think you may need to rebase the Easegress main first.

``` bash
megaease-km5:➜  easegress  git:(5cb82af*)|>make                                                                                          
[~/prtest/easegress]
build client
cd /home/ubuntu/prtest/easegress/ && \
CGO_ENABLED=0 go build -v -trimpath -ldflags "-s -w -X github.com/megaease/easegress/pkg/version.RELEASE=v1.1.0 -X github.com/megaease/easegress/pkg/version.COMMIT=git-5cb82af -X github.com/megaease/easegress/pkg/version.REPO=https://375cd9d91bd3efb3411b72037081bea5a65337aa@github.com/megaease/easegress.git" \
-o /home/ubuntu/prtest/easegress/bin/egctl /home/ubuntu/prtest/easegress/cmd/client
build server
cd /home/ubuntu/prtest/easegress/ && \
CGO_ENABLED=0 go build  -v -trimpath -ldflags "-s -w -X github.com/megaease/easegress/pkg/version.RELEASE=v1.1.0 -X github.com/megaease/easegress/pkg/version.COMMIT=git-5cb82af -X github.com/megaease/easegress/pkg/version.REPO=https://375cd9d91bd3efb3411b72037081bea5a65337aa@github.com/megaease/easegress.git" \
-o /home/ubuntu/prtest/easegress/bin/easegress-server /home/ubuntu/prtest/easegress/cmd/server
github.com/megaease/easegress/pkg/common
# github.com/megaease/easegress/pkg/common
pkg/common/time.go:24:1: syntax error: non-declaration statement outside function body
pkg/common/time.go:40:1: syntax error: non-declaration statement outside function body
Makefile:51: recipe for target 'build_server' failed
make: *** [build_server] Error 2

```

2. Easegress has introduced the `GitHub Actions / analysis` for static code analysis. This bot has commented in this PR too. Please take a look and fix them. Thx.

3. I think this PR is OK to me after all related problems are solved even without `graceful update`. So mabybe you can remove the `graceful update` related change in this PR too.

**Comment by tg123** [comment]
> > could this get merged to make windows supported in mainstream?
> > gracenet like cannot impl on windows at the moment.
> 
> Hi @tg123, first of all, thank you for providing a Windows supporting PR for Easegress.
> 
> 1. It seems that after v1.0.1 and v1.1.0, the Easegress' code implementation has evolved a lot, so the megaeasex bot can't build it in the test environment successfully. So I think you may need to rebase the Easegress main first.
> 
> ```shell
> megaease-km5:➜  easegress  git:(5cb82af*)|>make                                                                                          
> [~/prtest/easegress]
> build client
> cd /home/ubuntu/prtest/easegress/ && \
> CGO_ENABLED=0 go build -v -trimpath -ldflags "-s -w -X github.com/megaease/easegress/pkg/version.RELEASE=v1.1.0 -X github.com/megaease/easegress/pkg/version.COMMIT=git-5cb82af -X github.com/megaease/easegress/pkg/version.REPO=https://375cd9d91bd3efb3411b72037081bea5a65337aa@github.com/megaease/easegress.git" \
> -o /home/ubuntu/prtest/easegress/bin/egctl /home/ubuntu/prtest/easegress/cmd/client
> build server
> cd /home/ubuntu/prtest/easegress/ && \
> CGO_ENABLED=0 go build  -v -trimpath -ldflags "-s -w -X github.com/megaease/easegress/pkg/version.RELEASE=v1.1.0 -X github.com/megaease/easegress/pkg/version.COMMIT=git-5cb82af -X github.com/megaease/easegress/pkg/version.REPO=https://375cd9d91bd3efb3411b72037081bea5a65337aa@github.com/megaease/easegress.git" \
> -o /home/ubuntu/prtest/easegress/bin/easegress-server /home/ubuntu/prtest/easegress/cmd/server
> github.com/megaease/easegress/pkg/common
> # github.com/megaease/easegress/pkg/common
> pkg/common/time.go:24:1: syntax error: non-declaration statement outside function body
> pkg/common/time.go:40:1: syntax error: non-declaration statement outside function body
> Makefile:51: recipe for target 'build_server' failed
> make: *** [build_server] Error 2
> ```
> 
> 1. Easegress has introduced the `GitHub Actions / analysis` for static code analysis. This bot has commented in this PR too. Please take a look and fix them. Thx.
> 2. I think this PR is OK to me after all related problems are solved even without `graceful update`. So maybe you can remove the `graceful update` related change in this PR too.

my pr rebased with mainstream for sure, the commit is `d2e5887`, seems your is 5cb82af.
and the gh action is green. could you please show me why build failed?

**Comment by benja-wu** [comment]
@tg123  the error message has been provided in my last comment, which is 
``` shell

-o /home/ubuntu/prtest/easegress/bin/easegress-server /home/ubuntu/prtest/easegress/cmd/server
github.com/megaease/easegress/pkg/common
github.com/megaease/easegress/pkg/common
pkg/common/time.go:24:1: syntax error: non-declaration statement outside function body
pkg/common/time.go:40:1: syntax error: non-declaration statement outside function body
Makefile:51: recipe for target 'build_server' failed
make: *** [build_server] Error 2

```
It seems that your rebasing bring back the `time.go` into your branch, which is not needed in this PR since you had split it into `time_unix.go` and `time_win.go`

**Comment by tg123** [comment]
git status to check if your local branch is clean
time.go should not have line 40 after merged
https://github.com/tg123/easegress/blob/win/pkg/common/time.go

**Review by megaeasex — CHANGES_REQUESTED** [review]
[TASK:easegress-pr-test FAILED]megaease/easegress Pull Request 74 Deploy Test failed

**Comment by benja-wu** [comment]
> git status to check if your local branch is clean
> time.go should not have line 40 after merged
> https://github.com/tg123/easegress/blob/win/pkg/common/time.go

After rerunning the auto-test task and open the `verbose` option, I get this error message 

``` bash

Applying: first build on windows
Using index info to reconstruct a base tree...
M	pkg/common/time.go
M	pkg/graceupdate/graceupdate.go
Falling back to patching base and 3-way merge...
Auto-merging pkg/graceupdate/graceupdate_unix.go
Auto-merging pkg/common/time.go
CONFLICT (content): Merge conflict in pkg/common/time.go
error: Failed to merge in the changes.
Patch failed at 0001 first build on windows

```
This auto-test bot will try to merge the request PR into main by running 

``` bash
git fetch origin +refs/pull/74/head:pr_74; git checkout pr_74;git rebase main;git merge main;
```

The error message in the last comment might be misleading cause I manually rerun the `make` command.
And `time.go`'s content will become 
``` golang 
package common

import (
        "time"
)

<<<<<<< HEAD
// Now is the current time
func Now() time.Time {
        var tv syscall.Timeval
        syscall.Gettimeofday(&tv)
        return time.Unix(0, syscall.TimevalToNsec(tv))
}

// NowUnixNano is the current Unix Nano time
func NowUnixNano() int64 {
        var tv syscall.Timeval
        syscall.Gettimeofday(&tv)
        return syscall.TimevalToNsec(tv)
}

// Since returns the elapsed time.
=======
>>>>>>> first build on windows
func Since(t time.Time) time.Duration {
        return Now().Sub(t)
}

``` 
due to this failed three-way merged.

**Comment by tg123** [comment]
just git merge branch win, you do not have to run rebase

**Review by megaeasex — CHANGES_REQUESTED** [review]
[TASK:easegress-pr-test FAILED]megaease/easegress Pull Request 74 Deploy Test failed

**Comment by benja-wu** [comment]
> just git merge branch win, you do not have to run rebase

Emm... change to use rebasing only can't solve this problem in my env. Does it work in yours @tg123 ? 

``` bash
√ easegress $ git status
On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean
√ easegress $ git fetch origin +refs/pull/74/head:pr_74                                    
remote: Enumerating objects: 224, done.
remote: Counting objects: 100% (161/161), done.
remote: Compressing objects: 100% (41/41), done.
remote: Total 109 (delta 72), reused 95 (delta 58), pack-reused 0
Receiving objects: 100% (109/109), 15.10 KiB | 67.00 KiB/s, done.
Resolving deltas: 100% (72/72), completed with 28 local objects.
From https://github.com/megaease/easegress
 * [new ref]         refs/pull/74/head -> pr_74
√ easegress $ 
√ easegress $  git checkout pr_74                
Switched to branch 'pr_74'
√ easegress $ git rebase main    
Auto-merging pkg/graceupdate/graceupdate_unix.go
Auto-merging pkg/common/time.go
CONFLICT (content): Merge conflict in pkg/common/time.go
error: could not apply 5641c7b... first build on windows
Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply 5641c7b... first build on windows
```
This conflict remains.

**Review by megaeasex — CHANGES_REQUESTED** [review]
[TASK:easegress-pr-test FAILED]megaease/easegress Pull Request 74 Deploy Test failed

**Review by megaeasex — APPROVED** [review]
[TASK:easegress-pr-test SUCCESS]megaease/easegress Pull Request 74 Deploy Test Success

**Comment by tg123** [comment]
git squash for you
not sure why you are doing rebase, but you now you can rebase it. this is done by github nowadays.

File: cmd/server/main.go [hunk 1]
```diff
@@ -58,6 +56,24 @@ func main() {
...
+			logger.Errorf("failed to read pidfile: %v", err)
+			os.Exit(1)
+		}
+
+		if err := common.SignalRaise(pid, common.SingalUsr2); err != nil {
+			logger.Errorf("failed to send signal: %v", err)
+			os.Exit(1)
+		}
+
+		logger.Infof("graceful uprade signal sent")
```
Comment (benja-wu) [inline cmd/server/main.go:72]:
```suggestion
		logger.Infof("graceful upgrade signal sent")
```

File: pkg/option/option.go [hunk 2]
```diff
@@ -95,6 +96,7 @@ func New() *Options {
 	opt.flags.BoolVarP(&opt.ShowConfig, "print-config", "c", false, "Print the configuration.")
 	opt.flags.StringVarP(&opt.ConfigFile, "config-file", "f", "", "Load server configuration from a file(yaml format), other command line flags will be ignored if specified.")
 	opt.flags.BoolVar(&opt.ForceNewCluster, "force-new-cluster", false, "Force to create a new one-member cluster.")
+	opt.flags.BoolVar(&opt.SignalUpgrade, "signal-upgrade", false, "Send upgrade signal to the pid and exit. The server will start a graceful upgrade after signal received.")
```
Comment (benja-wu) [inline pkg/option/option.go:99]:
```suggestion
	opt.flags.BoolVar(&opt.SignalUpgrade, "signal-upgrade", false, "Send an upgrade signal to the server based on the local pid file, then exit. The original server will start a graceful upgrade after signal received.")
```

File: cmd/server/main.go [hunk 3]
```diff
@@ -58,6 +56,24 @@ func main() {
 	defer logger.Sync()
 	logger.Infof("%s", version.Long)
 
+	if opt.SignalUpgrade {
```
Comment (benja-wu) [inline cmd/server/main.go:59]: So this option provides a manual way for sending the USR2 signal to the original Easegress process? It got me thinking first. Maybe we can leave some comments here.

**Review by benja-wu — COMMENTED** [review]
> git squash for you
> not sure why you are doing rebase, but you now you can rebase it. this is done by github nowadays.

Thanks.  Yeah,  I had changed it to use `merge` only. And the related auto-testing is successful now.

**Review by megaeasex — APPROVED** [review]
[TASK:easegress-pr-test SUCCESS]megaease/easegress Pull Request 74 Deploy Test Success

**Review by megaeasex — APPROVED** [review]
[TASK:easegress-pr-test SUCCESS]megaease/easegress Pull Request 74 Deploy Test Success

**Review by megaeasex — APPROVED** [review]
[TASK:easegress-pr-test SUCCESS]megaease/easegress Pull Request 74 Deploy Test Success

**Review by megaeasex — APPROVED** [review]
[TASK:easegress-pr-test SUCCESS]megaease/easegress Pull Request 74 Deploy Test Success

**Review by megaeasex — APPROVED** [review]
[TASK:easegress-pr-test SUCCESS]megaease/easegress Pull Request 74 Deploy Test Success

**Review by megaeasex — APPROVED** [review]
[TASK:easegress-pr-test SUCCESS]megaease/easegress Pull Request 74 Deploy Test Success

**Review by megaeasex — APPROVED** [review]
[TASK:easegress-pr-test SUCCESS]megaease/easegress Pull Request 74 Deploy Test Success

**Review by benja-wu — APPROVED** [review]
LGTM

**Review by xxx7xxxx — APPROVED** [review]

**Pull request merged**